### PR TITLE
adding early stop callback to ptuning

### PR DESCRIPTION
**Pull request by arendu**
# What does this PR do ?

Adds a early stopping callback to the ptl trainer object for prompt and p-tuning for gpt and t5 models.

**Collection**: [NLP]

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by arendu** [comment]
experiment w and w/o early stop: https://wandb.ai/nvidia/earlystop?workspace=user-adithyare

**Review by mbitaab — APPROVED** [review]
Looks great, thanks!

**Review by Davood-M — DISMISSED** [review]
Looks great, thanks!

File: examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml [hunk 1]
```diff
@@ -36,6 +37,13 @@ exp_manager:
     filename: 'megatron_gpt_prompt_tune--{val_loss:.3f}-{step}'
     model_parallel_size: ${model.tensor_model_parallel_size}
     save_best_model: True
+  early_stopping_callback_params:
```
Comment (aklife97) [inline examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:40]: can we move this out of `exp_manager` OR move the callback code inside exp_manager?

Comment (arendu) [reply examples/nlp/language_modeling/conf/megatron_gpt_prompt_learning_config.yaml:40]: (discussed offline) will update exp_manager core in a separate PR.

**Review by aklife97 — DISMISSED** [review]
LGTM!

File: nemo/utils/exp_manager.py [hunk 2]
```diff
@@ -69,6 +70,15 @@ class CheckpointMisconfigurationError(NeMoBaseException):
     """ Raised when a mismatch between trainer.callbacks and exp_manager occurs"""
 
 
+@dataclass
+class EarlyStoppingParams:
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:74]: Add test that checks the dataclass args against PTL EarlyStoppingCallback class. Look at [test inside tests/core/lightning](https://github.com/NVIDIA/NeMo/blob/main/tests/core/test_config_utils.py#L117)

File: nemo/utils/exp_manager.py [hunk 3]
```diff
@@ -272,6 +284,7 @@ def exp_manager(trainer: 'pytorch_lightning.Trainer', cfg: Optional[Union[DictCo
                 pytorch lightning trainer. The ModelCheckpoint saves the top 3 models with the best "val_loss", the most
                 recent checkpoint under ``*last.ckpt``, and the final checkpoint after training completes under ``*end.ckpt``.
                 Defaults to True.
+            - create_early_stopping_callback (bool): Flag to decide if early stopping should be used to stop training. Default is False. See EarlyStoppingParams dataclass above.
```
Comment (titu1994) [inline nemo/utils/exp_manager.py:287]: New line after False.

File: examples/nlp/language_modeling/conf/megatron_t5_prompt_learning.yaml [hunk 4]
```diff
@@ -34,6 +34,12 @@ exp_manager:
     filename: "megatron_t5_prompt_tune--{${exp_manager.checkpoint_callback_params.monitor}:.3f}-{step}"
     model_parallel_size: ${model.tensor_model_parallel_size}
     save_best_model: True
+  early_stopping_callback_params:
```
Comment (titu1994) [inline examples/nlp/language_modeling/conf/megatron_t5_prompt_learning.yaml:37]: Add `create_early_stopping_callback` arg

File: tests/core/test_config_utils.py [hunk 5]
```diff
@@ -126,3 +128,12 @@ def test_ptl_config(self):
         assert signatures_match
         assert cls_subset is None
         assert dataclass_subset is None
+
+    @pytest.mark.unit
+    def test_early_stopping_config(self,):
```
Comment (arendu) [inline tests/core/test_config_utils.py:133]: @titu1994 I can't figure out what is wrong here.

Comment (titu1994) [reply tests/core/test_config_utils.py:133]: Seems to have worked out

**Review by titu1994 — APPROVED** [review]
Looks great !

**Pull request merged**